controllers/machine_controller_noderef.go

```diff
@@ -163,30 +164,17 @@ func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string)
 }
 
 func (r *MachineReconciler) getNode(ctx context.Context, c client.Reader, providerID *noderefutil.ProviderID) (*corev1.Node, error) {
-	log := ctrl.LoggerFrom(ctx, "providerID", providerID)
-
 	nodeList := corev1.NodeList{}
-	for {
-		if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
-			return nil, err
-		}
-
-		for _, node := range nodeList.Items {
-			nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
-			if err != nil {
-				log.Error(err, "Failed to parse ProviderID", "node", node.Name)
-				continue
-			}
-
-			if providerID.Equals(nodeProviderID) {
-				return &node, nil
-			}
-		}
+	if err := c.List(ctx, &nodeList, client.MatchingFields{remote.NodeProviderIDIndex: providerID.IndexKey()}); err != nil {
+		return nil, err
```

Comment:
Just a note that field indexer don't work with the fake client (which we're trying not to use throughout CAPI, but we should make sure there are no tests with the fake client that use this.